sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java

```diff
@@ -148,13 +148,17 @@ private void verifyUnmodifiedThrowingCheckedExceptions() throws CoderException {
       // encodedObject, because the Coder may treat it differently.
       //
       // For example, an unbounded Iterable will be encoded in an unbounded way, but decoded into an
-      // ArrayList, which will then be re-encoded in a bounded format. So we really do need to
-      // encode-decode-encode retainedObject.
-      if (Arrays.equals(
-          CoderUtils.encodeToByteArray(coder, clonedOriginalObject),
-          CoderUtils.encodeToByteArray(coder, clonedPossiblyModifiedObject))) {
-        return;
-      }
+      // ArrayList, which will then be re-encoded in a bounded format. So we get a structural value
+      // from a coder and used that to check if the byte array[] are the same. The structuralValue()
+      // method of the Coder returns a StructuralByteArray object.
```

Comment:
The jira issue could be worded better but my intent was we should rely only on the structural value comparison and drop all the other checks which are occurring within this method. Contractually the coder is required to return a value with a good equals method (not needing the usage of deep equals or cloning).